Initial Starlight Plugins support (#942)
withastro/starlight: docs/src/content/docs/reference/plugins.md

### `hooks.setup`

Plugin setup function called when Starlight is initialized (during the [`astro:config:setup`](https://docs.astro.build/en/reference/integrations-reference/#astroconfigsetup) integration hook).
The `setup` hook can be used to update the Starlight configuration or add Astro integrations.

This hook is called with the following options:

#### `config`

**type:** `StarlightUserConfig`

A read-only copy of the user-supplied [Starlight configuration](/reference/configuration).
This configuration may have been updated by other plugins configured before the current one.

#### `updateConfig`

**type:** `(newConfig: StarlightUserConfig) => void`

A callback function to update the user-supplied [Starlight configuration](/reference/configuration).
You only need to provide the root-level configuration keys that you want to update but no deep merge is performed. In order to update nested configuration values, you must provide the entire nested object.

For example, to add a new [`social`](/reference/configuration/#social) media account to the configuration without overriding the existing ones:

> ```ts {5-10}
> // plugin.ts
> export default {
>   name: 'add-twitter-plugin',
>   plugin({ config, updateConfig }) {
>     updateConfig({
>       social: {
>         ...config.social,
>         twitter: 'astrodotbuild',
>       },
>     });
>   },

Reviewer: Totally missed this on my first read through, but just realising the examples all still use the old shape 😅
````suggestion
```ts {6-11}
// plugin.ts
export default {
  name: 'add-twitter-plugin',
  hooks: {
    setup({ config, updateConfig }) {
      updateConfig({
        social: {
          ...config.social,
          twitter: 'astrodotbuild',
        },
      });
    },
  },
````

Author: Damn, good catch, can't believe I missed that so many times too 😨

};
```

#### `addIntegration`

**type:** `(integration: AstroIntegration) => void`

A callback function to add an [Astro integration](https://docs.astro.build/en/reference/integrations-reference/) required by the plugin.
For example, to add the [React Astro integration](https://docs.astro.build/en/guides/integrations-guide/react/):

> ```ts {13}
> // plugin.ts
> import react from '@astrojs/react';
> 
> export default {
>   name: 'plugin-using-react',
>   plugin({ addIntegration, astroConfig }) {
>     const isReactLoaded = astroConfig.integrations.find(
>       ({ name }) => name === '@astrojs/react'
>     );
> 
>     // Only add the React integration if it's not already loaded.
>     if (!isReactLoaded) {
>       addIntegration(react());
>     }
>   },

Reviewer: ````suggestion
```ts {14}
// plugin.ts
import react from '@astrojs/react';

export default {
  name: 'plugin-using-react',
  hooks: {
    plugin({ addIntegration, astroConfig }) {
      const isReactLoaded = astroConfig.integrations.find(
        ({ name }) => name === '@astrojs/react'
      );

      // Only add the React integration if it's not already loaded.
      if (!isReactLoaded) {
        addIntegration(react());
      }
    },
  },
````

};
```

#### `astroConfig`

**type:** `AstroConfig`

A read-only copy of the user-supplied [Astro configuration](https://docs.astro.build/en/reference/configuration-reference/).
This configuration is resolved before any other integrations have run.

#### `command`

**type:** `'dev' | 'build' | 'preview'`

The command used to run Starlight:

- `dev` - Project is executed with `astro dev`
- `build` - Project is executed with `astro build`
- `preview` - Project is executed with `astro preview`

#### `isRestart`

**type:** `boolean`

`false` when the dev server starts, `true` when a reload is triggered.
Common reasons for a restart include a user editing their `astro.config.mjs` while the dev server is running.

#### `logger`

**type:** `AstroIntegrationLogger`

An instance of the [Astro integration logger](https://docs.astro.build/en/reference/integrations-reference/#astrointegrationlogger) that you can use to write logs.
All logged messages will be prefixed with the plugin name.

> ```ts {5}
> // plugin.ts
> export default {
>   name: 'long-process-plugin',
>   plugin({ logger }) {
>     logger.info('Starting long process…');
>     // Some long process…
>   },

Reviewer: ````suggestion
```ts {6}
// plugin.ts
export default {
  name: 'long-process-plugin',
  hooks: {
    plugin({ logger }) {
      logger.info('Starting long process…');
      // Some long process…
    },
  },
````

A read-only copy of the user-supplied [Starlight configuration](/reference/configuration).
This configuration may have been updated by other plugins configured before the current one.

#### `updateConfig`

**type:** `(newConfig: StarlightUserConfig) => void`

A callback function to update the user-supplied [Starlight configuration](/reference/configuration).
You only need to provide the root-level configuration keys that you want to update but no deep merge is performed. In order to update nested configuration values, you must provide the entire nested object.

For example, to add a new [`social`](/reference/configuration/#social) media account to the configuration without overriding the existing ones:

```ts {6-11}
// plugin.ts
export default {
  name: 'add-twitter-plugin',
  hooks: {
    setup({ config, updateConfig }) {
      updateConfig({
        social: {
          ...config.social,
          twitter: 'astrodotbuild',
        },
      });
    },
  },
};
```

#### `addIntegration`

**type:** `(integration: AstroIntegration) => void`

A callback function to add an [Astro integration](https://docs.astro.build/en/reference/integrations-reference/) required by the plugin.

> For example, to add the [React Astro integration](https://docs.astro.build/en/guides/integrations-guide/react/):

Reviewer: Small suggestion to add the context around first checking before registering a new integration (which I think is a good practice for most integrations):

```suggestion

In the following example, the plugin first checks if [Astro’s React integration](https://docs.astro.build/en/guides/integrations-guide/react/) is configured and, if it isn’t, uses `addIntegration()` to add it:
```

> ```ts {14}

Reviewer: Maybe makes sense to also highlight the `addIntegration` in the plugin options?

````suggestion
```ts {14} "addIntegration,"
````

// plugin.ts
import react from '@astrojs/react';

export default {
  name: 'plugin-using-react',
  hooks: {
    plugin({ addIntegration, astroConfig }) {
      const isReactLoaded = astroConfig.integrations.find(
        ({ name }) => name === '@astrojs/react'
      );

      // Only add the React integration if it's not already loaded.
      if (!isReactLoaded) {
        addIntegration(react());
      }
    },
  },
};
```

#### `astroConfig`

**type:** `AstroConfig`

A read-only copy of the user-supplied [Astro configuration](https://docs.astro.build/en/reference/configuration-reference/).

> This configuration is resolved before any other integrations have run.

Reviewer: Is this 100% accurate? I’m pretty sure Starlight receives the config as transformed by any preceding integrations. If it _is_ accurate, maybe should be “Astro integrations” for clarity.

Author: Good question, this is totally based on https://docs.astro.build/en/reference/integrations-reference/#config-option :thinking:

Author: Should we just remove the sentence then?

PS: not sure why everything is red now, looks like it's related to the size check, I'll try to see why 🤔

Reviewer: Yes, I think removing the sentence is fine.

Yeah, not sure why either — looks like the docs site is failing to build all of a sudden.

---
title: Plugins Reference
description: An overview of the Starlight plugin API.
---

Starlight Plugins provide an API to customize Starlight configuration, UI and behavior while also being easy to share and reuse.
This page is for anyone interested in creating their own Starlight plugins.

Lean more about using a Starlight plugin in the [Configuration Reference](/reference/configuration/#plugins) or visit the [Starlight Plugins Showcase](/showcase/#community-plugins) to see a list of community plugins.

## Quick API Reference

```ts
interface StarlightPlugin {
  name: string;
  plugin: (options: {
    config: StarlightUserConfig;
    updateConfig: (newConfig: StarlightUserConfig) => void;
    addIntegration: (integration: AstroIntegration) => void;
    logger: AstroIntegrationLogger;
  }) => void | Promise<void>;

> }

Author: I was never a huge fan of the `plugin()` method name but did not managed at the time to come up with a better name I liked. I personally like `setup()` so I'm all for it.

Regarding the structure, I personally do not have a strong preference, so I think if you have one, we should go with it otherwise I guess mirroring Astro's structure may be the safest bet?

> }

Reviewer: No strong preference from me, so yeah, +1 to matching Astro’s shape 👍
